src/main/java/de/tum/in/www1/artemis/service/ExamService.java

```diff
@@ -115,4 +140,163 @@ public void delete(Long examId) {
+     */
+    public List<StudentDTO> registerStudentsForExam(@PathVariable Long courseId, @PathVariable Long examId, @RequestBody List<StudentDTO> studentDtos) {
+        var course = courseService.findOne(courseId);
+        var exam = findOneWithRegisteredUsers(examId);
+        List<StudentDTO> notFoundStudentsDtos = new ArrayList<>();
+        for (var studentDto : studentDtos) {
+            var registrationNumber = studentDto.getRegistrationNumber();
+            try {
+                // 1) we use the registration number and try to find the student in the Artemis user database
+                Optional<User> optionalStudent = userService.findUserWithGroupsAndAuthoritiesByRegistrationNumber(registrationNumber);
+                if (optionalStudent.isPresent()) {
+                    var student = optionalStudent.get();
+                    // we only need to add the student to the course group, if the student is not yet part of it, otherwise the student cannot access the exam (within the course)
+                    if (!student.getGroups().contains(course.getStudentGroupName())) {
+                        userService.addUserToGroup(student, course.getStudentGroupName());
+                    }
+                    exam.addUser(student);
+                    continue;
+                }
+                // 2) if we cannot find the student, we use the registration number and try to find the student in the (TUM) LDAP, create it in the Artemis DB and in a potential
+                // external user management system
+                optionalStudent = userService.createUserFromLdap(registrationNumber);
+                if (optionalStudent.isPresent()) {
+                    var student = optionalStudent.get();
+                    // the newly created student needs to get the rights to access the course, otherwise the student cannot access the exam (within the course)
+                    userService.addUserToGroup(student, course.getStudentGroupName());
+                    exam.addUser(student);
+                    continue;
+                }
+                // 3) if we cannot find the user in the (TUM) LDAP, we report this to the client
+                log.warn("User with registration number " + registrationNumber + " not found in Artemis user database and not found in (TUM) LDAP");
+            }
+            catch (Exception ex) {
+                log.warn("Error while processing user with registration number " + registrationNumber + ": " + ex.getMessage(), ex);
+            }
+
+            notFoundStudentsDtos.add(studentDto);
+        }
+        examRepository.save(exam);
+        return notFoundStudentsDtos;
+    }
+
+    private List<Integer> assembleIndicesListWithRandomSelection(List<Integer> mandatoryIndices, List<Integer> optionalIndices, Long numberOfOptionalExercises) {
+        // Add all mandatory indices
+        List<Integer> indices = mandatoryIndices;
```

Comment:
I like that you explicitly assign this, it helps with understanding what you are doing here. I would suggest to copy the whole array instead of just referencing it. This way no one will use a modified mandatoryIndices in the future.

Reply:
Thanks for the comment! I've applied it here: https://github.com/ls1intum/Artemis/pull/1678/commits/c1b337e9fb26fccaef7ddc1e1e800c86ccefd4e9